linter: typescript/no-unnecessary-type-arguments autofix removes default generic required for constructor inference

### What version of Oxlint are you using?

1.68.0

### What command did you run?

pnpm exec oxlint --fix repro.ts

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```ts
import { defineConfig } from 'oxlint';

export default defineConfig({
  options: {
    typeAware: true,
  },
  rules: {
    'typescript/no-unnecessary-type-arguments': 'error',
  },
});
```


### What happened?

`typescript/no-unnecessary-type-arguments` reports and autofixes a default generic type argument on a constructor call:

```ts
const context = new RequestContext<unknown>([
  ['instructions', 'Reply for the agent.'],
  ['tools', {}],
]);
```
The fixer changes it to:

```ts
const context = new RequestContext([
  ['instructions', 'Reply for the agent.'],
  ['tools', {}],
]);
```
After the fix, TypeScript fails because the constructor infers a narrower RequestContext<{ ... }> instead of the intended untyped/default RequestContext<unknown>.

### Minimal repro
```ts
type RecordToTuple<T> = {
  [K in keyof T]: [K, T[K]];
}[keyof T][];

class RequestContext<
  Values extends Record<string, unknown> | unknown = unknown,
> {
  constructor(
    _iterable?: Values extends Record<string, unknown>
      ? RecordToTuple<Partial<Values>>
      : Iterable<readonly [string, unknown]>,
  ) {}

  set<K extends Values extends Record<string, unknown> ? keyof Values : string>(
    _key: K,
    _value: Values extends Record<string, unknown>
      ? K extends keyof Values
        ? Values[K]
        : never
      : unknown,
  ): void {}
}

type UntypedRequestContext = RequestContext;

function acceptsUntypedContext(_context: UntypedRequestContext): void {}

const context = new RequestContext<unknown>([
  ['instructions', 'Reply for the agent.'],
  ['tools', {}],
]);

acceptsUntypedContext(context);
```

### Expected behavior
The rule should not report/fix this type argument, because removing it changes the inferred constructor type and breaks downstream assignability.

`tsc --noEmit` passes before the fix.

### Actual behavior
Oxlint removes `<unknown>`.

Then `tsc --noEmit` fails:
```
error TS2345: Argument of type 'RequestContext<{ [x: string]: {}; } & ...>'
is not assignable to parameter of type 'UntypedRequestContext'.
  Type 'unknown' is not assignable to type '{ [x: string]: {}; } & ...'
```

### Notes
This appears related to the remaining default-type-argument path after inference reporting was reverted in oxc-project/tsgolint#892.

In this case the explicit type argument equals the default, but omitting it does not preserve semantics because constructor inference chooses a narrower type from the argument list.

--- a/internal/rules/no_unnecessary_type_arguments/no_unnecessary_type_arguments.go
+++ b/internal/rules/no_unnecessary_type_arguments/no_unnecessary_type_arguments.go
@@ -120,7 +120,67 @@ var NoUnnecessaryTypeArgumentsRule = rule.Rule{
 			return nil
 		}
 
-		checkArgsAndParameters := func(arguments *ast.NodeList, parameters []*ast.Node) {
+		typeNodeReferencesTypeParameter := func(typeNode *ast.Node, typeParameterSymbols map[*ast.Symbol]struct{}) bool {
+			var visit func(node *ast.Node) bool
+			visit = func(node *ast.Node) bool {
+				if node == nil {
+					return false
+				}
+
+				if ast.IsIdentifier(node) {
+					if symbol := ctx.TypeChecker.GetSymbolAtLocation(node); symbol != nil {
+						if _, ok := typeParameterSymbols[symbol]; ok {
+							return true
+						}
+					}
+				}
+
+				return node.ForEachChild(func(child *ast.Node) bool {
+					return visit(child)
+				})
+			}
+
+			return visit(typeNode)
+		}
+
+		constructorArgumentsCanInferTypeParameters := func(node *ast.Node, parameters []*ast.Node) bool {
+			if !ast.IsNewExpression(node) || len(node.Arguments()) == 0 {
+				return false
+			}
+
+			typeParameterSymbols := make(map[*ast.Symbol]struct{}, len(parameters))
+			for _, parameter := range parameters {
+				name := parameter.Name()
+				if name == nil {
+					continue
+				}
+				if symbol := ctx.TypeChecker.GetSymbolAtLocation(name); symbol != nil {
+					typeParameterSymbols[symbol] = struct{}{}
+				}
+			}
+			if len(typeParameterSymbols) == 0 {
+				return false
+			}
+
+			signature := checker.Checker_getResolvedSignature(ctx.TypeChecker, node, nil, checker.CheckModeNormal)
+			if signature == nil {
+				return false
+			}
+			declaration := checker.Signature_declaration(signature)
+			if declaration == nil || declaration.FunctionLikeData() == nil {
+				return false
+			}
+
+			for _, parameter := range declaration.Parameters() {
+				if typeNode := parameter.Type(); typeNode != nil && typeNodeReferencesTypeParameter(typeNode, typeParameterSymbols) {
+					return true
+				}
+			}
+
+			return false
+		}
+
+		checkArgsAndParameters := func(node *ast.Node, arguments *ast.NodeList, parameters []*ast.Node) {
 			if arguments == nil || parameters == nil || len(arguments.Nodes) == 0 || len(parameters) == 0 {
 				return
 			}
@@ -171,6 +231,12 @@ var NoUnnecessaryTypeArgumentsRule = rule.Rule{
 				return
 			}
 
+			// Removing the entire type argument list from a constructor can re-enable
+			// inference for all class type parameters used by constructor parameters.
+			if lastParamIndex == 0 && constructorArgumentsCanInferTypeParameters(node, parameters) {
+				return
+			}
+
 			ctx.ReportNodeWithFixes(typeArgument, buildUnnecessaryTypeParameterMessage(), func() []rule.RuleFix {
 				var removeRange core.TextRange
 				if lastParamIndex == 0 {
@@ -185,32 +251,32 @@ var NoUnnecessaryTypeArgumentsRule = rule.Rule{
 		return rule.RuleListeners{
 			ast.KindExpressionWithTypeArguments: func(node *ast.Node) {
 				expr := node.AsExpressionWithTypeArguments()
-				checkArgsAndParameters(expr.TypeArguments, getTypeParametersFromType(node, expr.Expression))
+				checkArgsAndParameters(node, expr.TypeArguments, getTypeParametersFromType(node, expr.Expression))
 			},
 			ast.KindTypeReference: func(node *ast.Node) {
 				expr := node.AsTypeReferenceNode()
-				checkArgsAndParameters(expr.TypeArguments, getTypeParametersFromType(node, expr.TypeName))
+				checkArgsAndParameters(node, expr.TypeArguments, getTypeParametersFromType(node, expr.TypeName))
 			},
 
 			ast.KindCallExpression: func(node *ast.Node) {
 				expr := node.AsCallExpression()
-				checkArgsAndParameters(expr.TypeArguments, getTypeParametersFromCall(node))
+				checkArgsAndParameters(node, expr.TypeArguments, getTypeParametersFromCall(node))
 			},
 			ast.KindNewExpression: func(node *ast.Node) {
 				expr := node.AsNewExpression()
-				checkArgsAndParameters(expr.TypeArguments, getTypeParametersFromCall(node))
+				checkArgsAndParameters(node, expr.TypeArguments, getTypeParametersFromCall(node))
 			},
 			ast.KindTaggedTemplateExpression: func(node *ast.Node) {
 				expr := node.AsTaggedTemplateExpression()
-				checkArgsAndParameters(expr.TypeArguments, getTypeParametersFromCall(node))
+				checkArgsAndParameters(node, expr.TypeArguments, getTypeParametersFromCall(node))
 			},
 			ast.KindJsxOpeningElement: func(node *ast.Node) {
 				expr := node.AsJsxOpeningElement()
-				checkArgsAndParameters(expr.TypeArguments, getTypeParametersFromCall(node))
+				checkArgsAndParameters(node, expr.TypeArguments, getTypeParametersFromCall(node))
 			},
 			ast.KindJsxSelfClosingElement: func(node *ast.Node) {
 				expr := node.AsJsxSelfClosingElement()
-				checkArgsAndParameters(expr.TypeArguments, getTypeParametersFromCall(node))
+				checkArgsAndParameters(node, expr.TypeArguments, getTypeParametersFromCall(node))
 			},
 		}
 	},

--- a/internal/rules/no_unnecessary_type_arguments/no_unnecessary_type_arguments_test.go
+++ b/internal/rules/no_unnecessary_type_arguments/no_unnecessary_type_arguments_test.go
@@ -439,6 +439,42 @@ async function useAppSession(event: H3Event) {
   return useSession<SessionData>(event, config);
 }
     `},
+		// https://github.com/oxc-project/oxc/issues/22915
+		{Code: `
+type RecordToTuple<T> = {
+  [K in keyof T]: [K, T[K]];
+}[keyof T][];
+
+class RequestContext<
+  Values extends Record<string, unknown> | unknown = unknown,
+> {
+  constructor(
+    _iterable?: Values extends Record<string, unknown>
+      ? RecordToTuple<Partial<Values>>
+      : Iterable<readonly [string, unknown]>,
+  ) {}
+
+  set<K extends Values extends Record<string, unknown> ? keyof Values : string>(
+    _key: K,
+    _value: Values extends Record<string, unknown>
+      ? K extends keyof Values
+        ? Values[K]
+        : never
+      : unknown,
+  ): void {}
+}
+
+type UntypedRequestContext = RequestContext;
+
+function acceptsUntypedContext(_context: UntypedRequestContext): void {}
+
+const context = new RequestContext<unknown>([
+  ['instructions', 'Reply for the agent.'],
+  ['tools', {}],
+]);
+
+acceptsUntypedContext(context);
+    `},
 	}, []rule_tester.InvalidTestCase{
 		{
 			Code: `
@@ -967,6 +1003,27 @@ type Default = Record<string, any>;
 type Alias = Default;
 function f<T = Default>() {}
 f();
+      `,
+			},
+			Errors: []rule_tester.InvalidTestCaseError{
+				{
+					Line:      5,
+					MessageId: "unnecessaryTypeParameter",
+				},
+			},
+		},
+		{
+			Code: `
+class C<T, U = unknown> {
+  constructor(_value: U) {}
+}
+new C<string, unknown>('value');
+      `,
+			Output: []string{`
+class C<T, U = unknown> {
+  constructor(_value: U) {}
+}
+new C<string>('value');
       `,
 			},
 			Errors: []rule_tester.InvalidTestCaseError{

